File: ios/MyIdModule.swift
//
//  MyIdModule.swift
//  ZeroX
//
//  Created by AlisherRakhimov on 06/10/2025.
//

import Foundation
import MyIdSDK
import React

@objc(MyIdModule)
class MyIdModule: RCTEventEmitter {
  @objc
  static func constantsToExport() -> [String: Any] {
    return ["initialCount": 0]
  }

  @objc
  override static func requiresMainQueueSetup() -> Bool {
    return true
  }

  @objc
  override func supportedEvents() -> [String]! {
    return ["onSuccess", "onError", "onUserExited"]
  }

  @objc
  func startMyId() {
    DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
      
      let config = MyIdConfig()
      config.clientId = "zerox_sdk-FQEPbkzGnIZrejYj6AHdXRsWIywt9lcRiDZDhXJC"
      config.clientHash = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsw3Ad+h8EgEjt+5sdTxveshhapa+Q0anEajGtEGt6KLJgOfk54AU/RwBIvBPFJRUQqOAbngtFFS6SCWt26AtG8QtRRVL+xWF//2u/66bXVjrHlCKuBQNVoISJ+YyfVLpOhQYlrRyLP23sKrJdB2PBYlovP1HCWFP56KUn5T1dSluBy5h81ZSfmsUJO5U1lKLli2WMOPCFl9K1/6TOuRSv70U/nZX+pRLCIPzrdlf9zCLL49OShztalJOYtXibasqTrNCd0sBzTNbiQ3uGkmK5RH+L2hi4dy1vDEwH7VqMLcogJXnTEYAZ3KCAxmIUXvkhDstWK5uH8Ru0uZskcR5GwIDAQAB"
      config.clientHashId = "7b4507ca-9b70-4e92-8bfe-767db25a0be2"
      config.environment = .production
      config.cameraShape = .circle
      config.locale = .uzbek
      MyIdClient.start(withConfig: config, withDelegate: self)
    }
  }
}

extension MyIdModule: MyIdClientDelegate {
  func onSuccess(result: MyIdSDK.MyIdResult) {
    if let image = result.image {
      if let imageData = image.jpegData(compressionQuality: 1) { // Compression quality can be adjusted
        let base64String = imageData.base64EncodedString(options: .lineLength64Characters)
        sendEvent(withName: "onSuccess", body: ["code": result.code, "comparison": result.comparisonValue, "image": base64String])
      } else {
        // Handle failure to get image data
        sendEvent(withName: "onError", body: ["message": "Failed to convert image to Data", "code": 0])
      }
    } else {
      // Handle the case where there is no image
      sendEvent(withName: "onError", body: ["message": "No image available", "code": 0])
    }
  }

  func onError(exception: MyIdSDK.MyIdException) {
    sendEvent(
      withName: "onError",
      body: [
        "message": exception.message,
        "code": exception.code
      ]
    )
  }

  func onUserExited() {
    sendEvent(
      withName: "onUserExited",
      body: [
        "message": "exited"
      ]
    )
  }
}

File: android/app/src/main/java/com/zeroxuz/MyIdModule.kt
package com.zeroxuz

import android.app.Activity
import android.content.Intent
import android.graphics.Bitmap
import android.util.Base64
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import uz.myid.android.sdk.capture.MyIdClient
import uz.myid.android.sdk.capture.MyIdConfig
import uz.myid.android.sdk.capture.MyIdException
import uz.myid.android.sdk.capture.MyIdResult
import uz.myid.android.sdk.capture.MyIdResultListener
import uz.myid.android.sdk.capture.model.MyIdCameraShape
import uz.myid.android.sdk.capture.model.MyIdEnvironment
import uz.myid.android.sdk.capture.model.MyIdEvent
import uz.myid.android.sdk.capture.model.MyIdGraphicFieldType
import uz.myid.android.sdk.capture.model.MyIdLocale
import java.io.ByteArrayOutputStream

class MyIdModule(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext),
    LifecycleEventListener, ActivityEventListener, MyIdResultListener {
    private val myIdClient = MyIdClient()

    init {
        reactContext.addLifecycleEventListener(this)
        reactContext.addActivityEventListener(this)
    }

    override fun getName(): String {
        return "MyIdModule"
    }

    override fun onHostResume() {
    }

    override fun onHostPause() {
    }

    override fun onHostDestroy() {
    }

    @ReactMethod
    fun addListener(eventName: String?) {
    }

    @ReactMethod
    fun removeListeners(count: Int?) {
    }

    override fun onActivityResult(
        activity: Activity,
        requestCode: Int,
        resultCode: Int,
        intent: Intent?
    ) {
        myIdClient.handleActivityResult(resultCode, this)
    }

    override fun onNewIntent(intent: Intent) {
    }

    override fun onSuccess(result: MyIdResult) {
        val params = Arguments.createMap()
        val bitmap = result.getGraphicFieldImageByType(MyIdGraphicFieldType.FacePortrait)
        val base64Image = bitmap?.let { encodeToBase64(it) }
        params.putString("code", result.code)
        params.putString("image", base64Image)
        sendEvent("onSuccess", params)
    }

    override fun onError(e: MyIdException) {
        val params = Arguments.createMap()
        params.putString("message", e.message)
        params.putInt("code", e.code!!)

        sendEvent("onError", params)
    }

    override fun onEvent(event: MyIdEvent) {
//        TODO("Not yet implemented")
    }

    override fun onUserExited() {
        val params = Arguments.createMap()
        params.putString("message", "User exited")

        sendEvent("onUserExited", params)
    }

    @ReactMethod
    fun startMyId(lang: String) {

        val defaultLang = when (lang) {
            "uz" -> MyIdLocale.Uzbek
            "ru" -> MyIdLocale.Russian
            "en" -> MyIdLocale.English
            else -> MyIdLocale.Uzbek
        }

        val myIdConfig: MyIdConfig =
            MyIdConfig.Builder("zerox_sdk-FQEPbkzGnIZrejYj6AHdXRsWIywt9lcRiDZDhXJC")
                .withClientHash(
                    "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAsw3Ad+h8EgEjt+5sdTxveshhapa+Q0anEajGtEGt6KLJgOfk54AU/RwBIvBPFJRUQqOAbngtFFS6SCWt26AtG8QtRRVL+xWF//2u/66bXVjrHlCKuBQNVoISJ+YyfVLpOhQYlrRyLP23sKrJdB2PBYlovP1HCWFP56KUn5T1dSluBy5h81ZSfmsUJO5U1lKLli2WMOPCFl9K1/6TOuRSv70U/nZX+pRLCIPzrdlf9zCLL49OShztalJOYtXibasqTrNCd0sBzTNbiQ3uGkmK5RH+L2hi4dy1vDEwH7VqMLcogJXnTEYAZ3KCAxmIUXvkhDstWK5uH8Ru0uZskcR5GwIDAQAB",
                    "7b4507ca-9b70-4e92-8bfe-767db25a0be2",
                )
                .withLocale(defaultLang)
                .withEnvironment(MyIdEnvironment.Production)
                .withCameraShape(MyIdCameraShape.Circle)
                .build()

        if (reactContext.currentActivity != null) {
            myIdClient.startActivityForResult(
                reactContext.currentActivity!!, 1, myIdConfig, this
            )
        }
    }

    private fun sendEvent(
        eventName: String,
        params: WritableMap?
    ) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    private fun encodeToBase64(bitmap: Bitmap): String {
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.PNG, 100, outputStream)
        return Base64.encodeToString(outputStream.toByteArray(), Base64.DEFAULT)
    }

}


